packages/ui/src/CodeEditor.tsx

```diff
@@ -0,0 +1,194 @@
+  value?: string
+  className?: string
+  options?: CodeOptions
+  onChange?: OnChangeCallback
+  customExtensions?: Extension[]
+  innerRef?: MutableRefObject<EditorViewRef | null>
+}
+
+const DEFAULT_OPTIONS: CodeOptions = {
+  language: null,
+  lineWrapping: false,
+  lineNumbers: true,
+}
+
+export const CodeEditor: FC<CodeEditorProps> = ({ value, className, options = {}, onChange, customExtensions, innerRef }) => {
+  const parentRef = useRef<HTMLDivElement | null>(null)
+  const cm = useRef<EditorView | null>(null)
+  const optionsMemoized = useDeepCompareMemo(() => options, [options])
+
+  // use onChange via a Ref so we avoid unnecessary calls if the caller does not use useCallback on onChange
+  const onChangeRef = useRef<OnChangeCallback | undefined>(onChange)
+
+  // use this handle to access the cm element (which is a `view`)
+  useImperativeHandle(innerRef, () => ({
+    view: () => cm.current,
+  }))
+
+  useEffect(
+    () => {
+      if (!parentRef.current) return
+
+      const opts: CodeOptions = { ...DEFAULT_OPTIONS, ...optionsMemoized }
+
+      const updateListenerExtension = () =>
+        EditorView.updateListener.of((v: ViewUpdate) => {
+          if (!v.docChanged) return
+
+          const val = v.state.doc.toString()
+
+          const cb = onChangeRef.current
+          if (cb) cb(val)
+        })
+
+      // configure/enable very common features
+      const basicExtensions = [
+        highlightSpecialChars(),
+        history(),
+        drawSelection(),
+        EditorState.allowMultipleSelections.of(true),
+        indentOnInput(),
+        defaultHighlightStyle.fallback,
+        bracketMatching(),
+        closeBrackets(),
+        autocompletion(),
+        rectangularSelection(),
+        highlightActiveLine(),
+        highlightSelectionMatches(),
+        keymap.of([
+          ...closeBracketsKeymap,
+          ...defaultKeymap,
+          ...searchKeymap,
+          ...historyKeymap,
+          ...foldKeymap,
+          ...commentKeymap,
+          ...completionKeymap,
+          ...lintKeymap,
+        ]),
+      ]
+
+      const extensions = [
+        ...basicExtensions,
+        ...(customExtensions ?? []),
+        ...(opts.language ? [StreamLanguage.define(opts.language)] : []),
+        ...(opts.lineWrapping ? [EditorView.lineWrapping] : []),
+        ...(opts.lineNumbers ? [foldGutter(), lineNumbers()] : []),
+        updateListenerExtension(),
+      ]
+
+      cm.current = new EditorView({
+        state: EditorState.create({
+          doc: value || '',
+          extensions,
+        }),
+        parent: parentRef.current,
+      })
+
+      return () => {
+        // destroy cm. (mind you; this component will be recreated whenever options are modified.)
+        cm.current?.destroy()
+        cm.current = null
+      }
+    },
+    // value should not be a dependency (it is handled below), just disable the warnings:
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```

Comment:
Do we still need to disable the linter check on this line?

Reply:
@aigoncharov yes, for .value.